File: 2024/day6.py
from tqdm import tqdm


def gen_grid(string):
    return list(map(list, string.split()))


def to_string(grid):
    return "\n".join(map("".join, grid))


def test_data():
    return gen_grid(
        """....#.....
    .........#
    ..........
    ..#.......
    .......#..
    ..........
    .#..^.....
    ........#.
    #.........
    ......#..."""
    )


def real_data():
    from pathlib import Path

    return gen_grid(Path("input6.txt").read_text())


class Walker:
    def __init__(self, data: list[list[str]]):
        self.grid = data
        self.start = self.get_start()
        self.curr = self.start
        self.dxy = (-1, 0)
        self.seen: set[tuple[tuple[int, int], tuple[int, int]]] = set()

    def get_start(self) -> tuple[int, int]:
        for i, row in enumerate(self.grid):
            for j, char in enumerate(row):
                if char == "^":
                    return i, j
        assert False

    def step(self):
        x, y = self.curr
        dx, dy = self.dxy
        if self.grid[x][y] != "#":
            self.seen.add((self.dxy, self.curr))
            self.grid[x][y] = "X"
            self.curr = (x + dx, y + dy)
        else:
            self.curr = (x - dx, y - dy)
            self.dxy = (dy, -dx)

    def within_bounds(self):
        x, y = self.curr
        return 0 <= x < len(self.grid) and 0 <= y < len(self.grid[0])

    def simulate(self):
        while self.within_bounds():
            self.step()
            if (self.dxy, self.curr) in self.seen:
                return True
        return False

    def reset(self):
        self.seen.clear()
        self.curr = self.start
        self.dxy = (-1, 0)


def part1(wk: Walker) -> tuple[int, set[tuple[int, int]]]:
    from operator import itemgetter

    wk.reset()
    wk.simulate()
    return sum(x.count("X") for x in wk.grid), set(map(itemgetter(1), wk.seen))


def part2(wk: Walker, candidates: set[tuple[int, int]]) -> int:
    wk.reset()
    loops = 0
    for i, j in tqdm(candidates, leave=False):
        wk.grid[i][j] = "#"
        if wk.simulate():
            loops += 1
        wk.reset()
        wk.grid[i][j] = "."

    return loops


wk = Walker(real_data())
ans, tocheck = part1(wk)
print("Part 1:", ans)
print("Part 2:", part2(wk, tocheck))

File: 2024/day6.cpp
#include "day6-input.h"
#include <algorithm>
#include <cstdint>
#include <iostream>
#include <iterator>
#include <unordered_set>
#include <utility>
#include <vector>

using Grid = std::vector<std::vector<uint8_t>>;
using Coord = std::pair<int8_t, int8_t>;

#define PACKI4(dx, dy, x, y)                                                   \
  (uint32_t)(((uint8_t)dx) << 24 | ((uint8_t)dy) << 16 | ((uint8_t)x) << 8 |   \
             ((uint8_t)y))

class Walker {
public:
  Grid grid;
  int8_t sx, sy, x, y, dx, dy;
  std::unordered_set<uint32_t> seen = {};

  Walker(Grid &&grid) : grid(grid), dx(-1), dy(0) {
    for (int i = 0; i < grid.size(); i++) {
      for (int j = 0; j < grid.size(); j++) {
        if (grid[i][j] == '^') {
          sx = i;
          sy = j;

          x = i;
          y = j;
        }
      }
    }
  }

  void step() {
    if (grid[x][y] != '#') {
      seen.insert(PACKI4(dx, dy, x, y));
      grid[x][y] = 'X';
      x += dx;
      y += dy;
    } else {
      x -= dx;
      y -= dy;

      dx = -dx;
      std::swap(dx, dy);
    }
  }

  bool within_bounds() {
    return 0 <= x and 0 <= y and x < grid.size() and y < grid[0].size();
  }

  bool simulate() {
    while (within_bounds()) {
      step();
      if (seen.contains(PACKI4(dx, dy, x, y))) {
        return true;
      }
    }
    return false;
  }

  void reset() {
    seen.clear();
    x = sx;
    y = sy;
    dx = -1;
    dy = 0;
  }
};

std::pair<int, std::unordered_set<uint32_t>> part1(Walker &wk) {
  wk.reset();
  wk.simulate();
  int cover = 0;
  for (const auto &row : wk.grid) {
    cover += std::count(row.cbegin(), row.cend(), 'X');
  }
  return {cover, wk.seen};
}

int part2(Walker &wk, std::unordered_set<uint16_t> &candidates) {
  wk.reset();
  int loops = 0;
  for (const auto h : candidates) {
    int8_t i = h >> 8;
    int8_t j = h;
    wk.grid[i][j] = '#';
    if (wk.simulate()) {
      loops++;
    }
    wk.reset();
    wk.grid[i][j] = '.';
  }
  return loops;
}

int main() {
  auto wk = Walker(INPUT);
  auto [ans, visited] = part1(wk);
  std::unordered_set<uint16_t> tocheck;
  std::transform(visited.cbegin(), visited.cend(),
                 std::inserter(tocheck, tocheck.end()), std::identity{});
  std::cout << "Part 1: " << ans << std::endl;
  std::cout << "Part 2: " << part2(wk, tocheck) << std::endl;
}
